cookbook/case_studies/ml_training/mnist_classifier/pytorch_single_node_multi_gpu.py

```diff
@@ -0,0 +1,406 @@
+            images, targets = images.to(rank), targets.to(rank)  # device conversion
+            outputs = model(images)  # forward pass -- generate predictions
+            test_loss += F.nll_loss(outputs, targets, reduction="sum").item()  # sum up batch loss
+            _, predicted = torch.max(outputs.data, 1)  # get the index of the max log-probability
+            correct += (predicted == targets).sum().item()  # compare predictions to true label
+
+            # log predictions to the ``wandb`` table
+            if log_counter < NUM_BATCHES_TO_LOG:
+                if rank == 0:
+                    log_test_predictions(images, targets, outputs, predicted, my_table, log_counter)
+                log_counter += 1
+
+    # compute the average loss
+    test_loss /= total
+    accuracy = float(correct) / total
+
+    if rank == 0:
+        print("\ntest_loss={:.4f}\naccuracy={:.4f}\n".format(test_loss, accuracy))
+        # log the average loss, accuracy, and table
+        wandb.log({"test_loss": test_loss, "accuracy": accuracy, "mnist_predictions": my_table})
+
+    return accuracy
+
+
+# %%
+# Hyperparameters
+# ===============
+#
+# We define a few hyperparameters for training our model.
+@dataclass_json
+@dataclass
+class Hyperparameters(object):
+    """
+    Args:
+        backend: pytorch backend to use, e.g. "gloo" or "nccl"
+        sgd_momentum: SGD momentum (default: 0.5)
+        seed: random seed (default: 1)
+        log_interval: how many batches to wait before logging training status
+        batch_size: input batch size for training (default: 64)
+        test_batch_size: input batch size for testing (default: 1000)
+        epochs: number of epochs to train (default: 10)
+        learning_rate: learning rate (default: 0.01)
+    """
+
+    backend: str = dist.Backend.GLOO
+    sgd_momentum: float = 0.5
+    seed: int = 1
+    log_interval: int = 10
+    batch_size: int = 64
+    test_batch_size: int = 1000
+    epochs: int = 10
+    learning_rate: float = 0.01
+
+
+# %%
+# Training and Evaluating
+# =======================
+
+TrainingOutputs = typing.NamedTuple(
+    "TrainingOutputs",
+    epoch_accuracies=typing.List[float],
+    model_state=PythonPickledFile,
+)
+
+MODEL_FILE = "./mnist_cnn.pt"
+ACCURACIES_FILE = "./mnist_cnn_accuracies.json"
+
+
+def dist_setup(rank, world_size, backend):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "8888"
+    dist.init_process_group(backend, rank=rank, world_size=world_size)
+
+
+def train_mnist(rank: int, world_size: int, hp: Hyperparameters) -> TrainingOutputs:
+
+    # store the hyperparameters' config in ``wandb``
+    if rank == 0:
+        wandb_setup()
+        wandb.config.update(json.loads(hp.to_json()))
+        print("wandb config:", wandb.config)
+
+    # set random seed
+    torch.manual_seed(hp.seed)
+
+    use_cuda = torch.cuda.is_available()
+    print(f"Using distributed PyTorch with {hp.backend} backend")
```

Comment:
Can we use a logger to log the statements?

Reply:
why? doesn't seem necessary... unless we want to use `logger` in all our flytesnacks examples moving forward

Reply:
Okay.